loadgen/generic_executor.go

```diff
@@ -68,34 +71,52 @@ func (g *GenericExecutor) newRun(info ScenarioInfo) (*genericRun, error) {
 // iterations is reached.
 func (g *genericRun) Run(ctx context.Context) error {
 	ctx, cancel := context.WithCancel(ctx)
-	if g.config.Duration > 0 {
-		ctx, cancel = context.WithTimeout(ctx, g.config.Duration)
+	if g.config.Timeout > 0 {
+		g.logger.Debugf("Will timeout after %v", g.config.Timeout)
+		ctx, cancel = context.WithTimeout(ctx, g.config.Timeout)
 	}
 	defer cancel()
 
+	durationCtx := ctx
+	if g.config.Duration > 0 {
+		durationCtx, cancel = context.WithTimeout(ctx, g.config.Duration)
+		defer cancel()
+	}
+
 	startTime := time.Now()
 	var runErr error
 	doneCh := make(chan error)
 	var currentlyRunning int
-	waitOne := func() {
+	waitOne := func(respectDuration bool) {
+		durationDone := make(<-chan struct{})
+		if respectDuration {
+			durationDone = durationCtx.Done()
+		}
 		select {
 		case err := <-doneCh:
 			currentlyRunning--
 			if err != nil {
 				runErr = err
 			}
 		case <-ctx.Done():
+		case <-durationDone:
 		}
 	}
 
-	// Run all until we've gotten an error or reached iteration limit
-	for i := 0; runErr == nil && ctx.Err() == nil &&
+	if g.config.Iterations > 0 {
+		g.logger.Debugf("Will run %v iterations", g.config.Iterations)
+	} else {
+		g.logger.Debugf("Will start iterations for %v", g.config.Duration)
+	}
+
+	// Run all until we've gotten an error or reached iteration limit or timeout
+	for i := 0; runErr == nil && durationCtx.Err() == nil && ctx.Err() == nil &&
```

Comment:
durationCtx is either equal to ctx or a child of ctx, no need to check both.

Reply:
Ah, yeah, good point. I blindly applied Chad's suggestion.